docs: scaling task agents (#70)
encord-team/encord-agents: docs/task_agents/runner_intro.md

The `Runner` classes are the core components for building task agents in Encord.
They provide a simple interface for defining agent logic and moving tasks through the Encord project workflows.

The `Runner`s manage the execution of agent logic on tasks within specific workflow stages.
They are responsible for:

- Connect directly to your Encord project via the Encord [SDK](https://docs.encord.com/sdk-documentation/getting-started-sdk/installation-sdk){ target="\_blank", rel="noopener noreferrer" }
- Provide function decorators to associate your agent logic with workflow stages
- Manage retries and error handling
- Handle task fetching and updates
- Optimize performance through batched updates and data loading

In the following sections, we will go through the different components of the `Runner`s and how to use them.

## Stage Decorators

Imagine that you have a workflow with three stages: `start`, `agent`, and `complete`.

> The `Runner` objects from `encord_agents.tasks.runner` will allow you to define the logic for the purple stages.

Reviewer: ```suggestion
The `Runner` objects from `encord_agents.tasks.runner` allow you to define the logic for the purple stages.
```

Reviewer 2: ```suggestion
The `Runner` objects from `encord_agents.tasks.runner` allows you to define the logic for the purple stages.
```
Would be this instead

> In this case, the `Runner` will allow you to define the logic for the `Agent 1` stage.

Reviewer: ```suggestion
In this case, the `Runner` allows you to define the logic for the `Agent 1` stage.
```

                    }
                ]
            },
            {
                "uuid": "7e7598de-612c-40c4-ba08-5dfec8c3ae8f",
                "nodeType": "DONE",
                "title": "Complete",
                "nodeSubtype": "COMPLETED"
            }
        ],
    }
    ```

The `@runner.stage` decorator connects your functions to specific stages in your Encord workflow.
For the workflow above, you would define the logic for the `Agent 1` stage as follows:

```python
@runner.stage(stage = "Agent 1")
# or @runner.stage(stage = "6011c844-fb26-438b-b465-0b0825951015")
def my_agent(lr: LabelRowV2, ...) -> str | UUID | None:
    """
    Args:
        lr: Automatically injected via by the `Runner`
        ...: See the "Dependencies" section for examples of
             how to, e.g., inject assets, client metadata, and
             more.

    Returns:
        The name or UUID of the pathway where the task should go next,
        or None to leave the task in the current stage.
    """
    pass
```

The agent function is supposed to return where the task should go next.
This can be done by pathways names or `UUID`s. 

> If None is returned, the task will not move and the runner will pick up that task again in the future.

Reviewer: ```suggestion
If None is returned, the task is not moved and the runner picks up that task again at a later stage.
```

You can also define multiple stages in a single runner:

```python
@runner.stage("prelabel")
def prelabel_task(lr: LabelRowV2) -> str:
    # Add initial labels
    return "review"

@runner.stage("validate")
def validate_task(lr: LabelRowV2) -> str:
    # Validate labels
    return "complete"
```

> If you define multiple stages. Depending on which type of runner you use, the execution logic will differ.

Reviewer: ```suggestion
If you define multiple stages. Depending on which type of runner you use, the execution logic differs
```

That is, if you define a runner with two stages:

=== "Runner"
    ```python
    runner = Runner()

    @runner.stage("stage_1")
    def stage_1():
        return "next"

    @runner.stage("stage_2")
    def stage_2():
        return "next"
    ```

>     The `Runner` will execute the tasks in the order in which the stages were defined in the runner.

Reviewer: ```suggestion
    The `Runner` executes the tasks in the order in which the stages were defined in the runner.
```

>     That is, the tasks will be processed by `stage_1` first and then by `stage_2`.

Reviewer: ```suggestion
    That is, the tasks are processed in `stage_1` first and in `stage_2` next.
```

=== "QueueRunner"
    ```python
    runner = QueueRunner()

    @runner.stage("stage_1")
    def stage_1():
        return "next"

    @runner.stage("stage_2")
    def stage_2():
        return "next"
    ```

>     The `QueueRunner` will give you control over the task queues for each stage.

Reviewer: ```suggestion
    The `QueueRunner` gives you control over the task queues for each stage.
```

    Please refer to the [QueueRunner documentation](./queue_runner.md) for more information.

### Optional arguments

> When you wrap a function with the `@runner.stage(...)` wrapper, you can include a [`label_row_metadata_include_args: LabelRowMetadataIncludeArgs`](../reference/core.md#encord_agents.core.data_model.LabelRowMetadataIncludeArgs) argument which will be passed on to the Encord Project's [`list_label_row_v2` method](https://docs.encord.com/sdk-documentation/sdk-references/project#list-label-rows-v2){ target="\_blank", rel="noopener noreferrer" }. This is useful to, e.g., be able to _read_ the client metadata associated to a task.

Reviewer: ```suggestion
When you wrap a function with the `@runner.stage(...)` wrapper, you can include a [`label_row_metadata_include_args: LabelRowMetadataIncludeArgs`](../reference/core.md#encord_agents.core.data_model.LabelRowMetadataIncludeArgs) argument to be passed on to the Encord Project's [`list_label_row_v2` method](https://docs.encord.com/sdk-documentation/sdk-references/project#list-label-rows-v2){ target="\_blank", rel="noopener noreferrer" }. This is useful to, e.g., be able to _read_ the client metadata associated to a task.
```

In the example below, we show how to obtain both label rows from "twin projects" and a frame iterator for videos -- just by specifying that it's something that the agent function depends on.

```python
from typing_extensions import Annotated
from encord.workflow.stages.agent import AgentStage
from encord_agents.tasks import Depends
from encord_agents.tasks.dependencies import (
    Twin,              # Access a "twin" project's labels
    dep_twin_label_row,# Get label row from twin project
    dep_video_iterator # Iterate over video frames
)

@runner.stage("my_stage")
def my_agent(
    task: AgentTask,
    lr: LabelRowV2,
    twin: Annotated[Twin, Depends(dep_twin_label_row(twin_project_hash="..."))],
    frames: Annotated[Iterator[Frame], Depends(dep_video_iterator)]
) -> str:
    # Use the dependencies
    pass
```

In the function body above, the `task`, `lr`, `twin`, and `frames` variables will be automatically injected with the respective dependencies.
This gives you the flexibility to only inject the dependencies that you need and focus on the logic of your agent.

#### Annotations
There are three object types that you can get without any extensive type annotations.

> If you type __any__ parameter of your stage implementation, e.g., the `my_agent` function above, with either of `[AgentTask, Project, LabelRowV2]`, the function will be called with that type of object, matching the task at hand.

Reviewer: ```suggestion
If you type __any__ parameter of your stage implementation, e.g., the `my_agent` function above, with either of `[AgentTask, Project, LabelRowV2]`, the function is called with that type of object, matching the task at hand.
```

That is, if you do:

```python
from encord.project import Project
...

@runner.stage("your_stage_name")
def my_agent(project: Project):
    ...
```

> The `project` will be the [workflow project][docs-workflow-project]{ target="\_blank", rel="noopener noreferrer" } instance for the `project_hash` you specified when defining or executing the runner.

Reviewer: ```suggestion
The `project` is the [workflow project][docs-workflow-project]{ target="\_blank", rel="noopener noreferrer" } instance for the `project_hash` that was specified when defining or executing the runner.
```

> 
> @runner.stage("my_stage")
> def my_agent(
>     # Notice the call to the function                                  👇
>     dep: Annotated[MyDependency, Depends(my_dependency_with_default_arg())]  
> ):
>     ...
> ```
> 
> This will allow you to set default values when you call the dependency.

You can create your own dependencies that can also use nested dependencies like this:

```python
from encord.objects import LabelRowV2
from encord.storage import StorageItem

def my_custom_dependency(
    lr: LabelRowV2,
    storage_item: StorageItem = Depends(dep_storage_item)
) -> dict:
    """Custom dependencies can use LabelRowV2 and other dependencies"""
    return {
        "data_title": lr.data_title,
        "metadata": storage_item.client_metadata
    }

@runner.stage("my_stage")
def my_agent(
    metadata: Annotated[dict, Depends(my_custom_dependency)]
) -> str:
    # metadata is automatically injected
    return "next_stage"
```

## Running the runner

There are two different types of runners with different use-cases. They also have two slightly different execution interfaces. 
Please refer to the following pages for more information:

> 1. [`Runner`](./sequential_runner.md#running-agents):  This is a simple sequential runner that will run the agent functions one after the other. It's easier to debug and understand. Use this for simple workflows or for testing out functionality before you scale it with the `QueueRunner`.

Reviewer: ```suggestion
1. [`Runner`](./sequential_runner.md#running-agents):  This is a simple sequential runner to run the agent functions one after the other. It is easier to debug and understand. Use this for simple workflows or for testing out functionality before you scale it with the `QueueRunner`.
```

> 2. [`QueueRunner`](./queue_runner.md#running-agents): This is a more advanced runner that will allow you to run the agent functions in parallel. It's useful when you have a lot of tasks to process and you want to speed up the processing time via parallel execution.

Reviewer: ```suggestion
2. [`QueueRunner`](./queue_runner.md#running-agents): This is a more advanced runner that allows you to run the agent functions in parallel. It's useful when you have a lot of tasks to process and you want to speed up the processing time via parallel execution.
```
